### PR TITLE
Support querying null/non-null key using B+ Tree index

### PR DESCRIPTION
**Pull request by Moonlit-Sailor**
## What changes were proposed in this pull request?
Support query for isNull/isNotNull in B+ tree index:
`attribute is null` , `attribute is not null` now is supported by B+ tree index
## How was this patch tested?
unit test

### CODE REVIEW

**Comment by Moonlit-Sailor** [comment]
please help reviewing the `BTreeIndexRecordWriter` and `BTreeIndexRecordWriter` files,
since I worked on it based on the `ChunkedBuffer` version before you refactored to Guava cache version.
Then I merged with the newest version, I concern that if there are some underlying bugs
@linhong-intel 
thanks  : )

**Comment by yueawang** [comment]
Just one question, Does intervalRange (DummyStart, DummyEnd) include those NULL value??

**Comment by Moonlit-Sailor** [comment]
@yueawang      
> Does intervalRange (DummyStart, DummyEnd) include those NULL value?

nope, [DummyStart , DummyEnd] is all non-null keys

**Comment by Moonlit-Sailor** [comment]
B+Tree Index File Structure
\************************************************
\*　　　　　　　 **Version   info**　　　　　　
\*———————————————————
\* 　　　　　　　**Node Data**　　　　
\*———————————————————
\* 　Row ID (with Non-Null key) List　　　　
\* 　Row ID (with Null key) List　　　　　<=======
\*———————————————————
\*　　　　　　　 **Footer**　　　
\* 　Row Count with Non-Null Key
\* 　Row Count With Null Key　　　　　<======
\* 　Node Count 
\*　 Nodes Meta Data
\*　 Min/Max Key Data for each node
\*———————————————————
\* 　Row Id List Section Size
\* 　Footer Section Size
\***********************************************

**Comment by yueawang** [comment]
Thanks @Moonlit-Sailor , that's what I have expected.

**Comment by linhong-intel** [comment]
LGTM but need some test cases.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala [hunk 1]
```diff
@@ -125,9 +130,10 @@ private[index] case class BTreeIndexRecordWriter(
         BTreeNodeMetaData(rowCount, nodeBuf.length, nodeUniqueKeys.head, nodeUniqueKeys.last)
       }
     // Write Row Id List
-    fileWriter.writeRowIdList(serializeRowIdLists(uniqueKeys))
+    fileWriter.writeRowIdList(serializeRowIdLists(nonNullUniqueKeys ++ nullKeys))
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala]: combine 2 arrays will involve memory consumption, can we modify the interface to avoid that?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala [hunk 2]
```diff
@@ -139,11 +143,13 @@ private[index] case class BTreeIndexRecordReader(
    */
   private def findNodeIdx(candidate: InternalRow, isStart: Boolean): (Option[Int], Boolean) = {
     val idxOption = (0 until footer.getNodesCount).find { idx =>
-      rowOrdering(candidate, footer.getMaxValue(idx, schema), isStart) <= 0
+      footer.getRowCountOfNode(idx) > 0 && // ensure this node is not an empty node
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala]: A node has zero row count? can you give some details here?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/UnsafeIndexTree.scala [hunk 3]
```diff
@@ -210,15 +210,25 @@ private[oap] class CurrentKey(node: IndexNode, keyIdx: Int, valueIdx: Int, index
...
 }
 
-private [oap] class RangeInterval(s: Key, e: Key, includeStart: Boolean, includeEnd: Boolean)
-  extends Serializable {
+private [oap] class RangeInterval(
+    s: Key,
+    e: Key,
+    includeStart: Boolean,
+    includeEnd: Boolean,
+    isNull: Boolean) extends Serializable {
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/UnsafeIndexTree.scala]: In my understanding, isNull and s, e, includeStart, includeEnd is exclusive.
So it's better to have two kinds of RangeInterval. one is for range, one is for null.
Please think about if we can make another PR for this

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala (hunk 2 above)
Comment (Moonlit-Sailor) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala]: all the keys are null

**Comment by adrian-wang** [comment]
LGTM pending tests.
